File: javascript/non_modifying_sequence.js
var all_of = function(ary, cb) {
  for (var i = 0; i < ary.length; i++) {
    if (!cb(ary[i])) return false;
  }
  return true;
};

var any_of = function(ary, cb) {
  for (var i = 0; i < ary.length; i++) {
    if (cb(ary[i])) return true;
  }
  return false;
};

var none_of = function(ary, cb) {
  return !Array.prototype.some.call(ary, cb);
};

var for_each = function(collection, cb) {
  if (collection.hasOwnProperty('length')) {
    for (var i = 0; i < collection.length; i++) {
      cb(collection[i]);
    }
  }
  else {
    for (var j in collection) {
      cb(collection[j]);
    }
  }
};

var count_if = function(collection, cb) {
  var count = 0;
  if (collection.hasOwnProperty('length')) {
    Array.prototype.forEach.call(collection, function(e){
      if (cb(e)) 
        count += 1;
    });
  }
  else {
    for (var i in collection) {
      if (cb(collection[i]))
        count += 1;
    }
  }
  return count;
};

var count = function(collection, match) {
  return count_if(collection, function(e){
    return e === match;
  });
};

var mismatch = function(ary1, ary2) {
  for (var i = 0; i < ary1.length; i++) {
    if (ary1[i] !== ary2[i]) {
      return i;
    }
  }
};

var equal = function(collection1, collection2) {
  return JSON.stringify(collection1) === JSON.stringify(collection2);
};

// Only consider that collection isn't an Array
var find_if = function(collection, cb) {
  for (var i in collection) {
    if (cb(collection[i])) {
      return i;
    }
  }
};

var find_if_not = function(collection, cb) {
  for (var i in collection) {
    if (!cb(collection[i])) {
      return i;
    }
  }
};

var find = function(collection, match) {
  return find_if(collection, function(e){
    return e === match;
  });
};

var find_end = function(ary, sub) {
  var eq = function(a, b) { return a === b; };
  var meet = search(ary, sub, eq);
  var lastMet = -1;
  while (meet != -1) {
    lastMet += (meet + 1);
    ary = Array.prototype.slice.call(ary, meet + 1);
    meet = search(ary, sub, eq);
  }
  return lastMet;
};

var find_first_of = function(ary, sub) {
  var meet = -1;
  for (var i = 0; i < sub.length; i++) {
    if ((meet = Array.prototype.indexOf.call(ary, sub[i])) != -1) {
      return meet;
    }
  }
  return meet;
};

var search = function(ary, sub, cb) {
  if (sub.length === 0)
    return 0;
  var start = Array.prototype.indexOf.call(ary, sub[0]);
  var mismatch = false;
  while (start != -1) {
    for (var i = 0; i < sub.length; i++) {
      if (!cb(ary[start + i], sub[i])) {
        mismatch = true;
        break;
      }
    }
    if (!mismatch) {
      return start;
    }
    else {
      start = Array.prototype.indexOf.call(ary, sub[0], start + 1);
    }
  }
  return -1;
};

var search_n = function(ary, n, value, cb) {
  var sub = new Array(n);
  for (var i = 0; i < sub.length; i++) {
    sub[i] = value;
  }
  return search(ary, sub, cb);
};

module.exports = {
  all_of: all_of,
  any_of: any_of,
  none_of: none_of,
  for_each: for_each,
  count: count,
  count_if: count_if,
  mismatch: mismatch,
  equal: equal,
  find: find,
  find_if: find_if,
  find_if_not: find_if_not,
  find_end: find_end,
  find_first_of: find_first_of,
  search: search,
  search_n: search_n
};

File: java/AlgorithmAndDataStructure/src/main/java/algorithms/NonModifyingSequence.java
package algorithms;

import functionObjects.BinaryPredicate;
import functionObjects.UnaryOp;
import functionObjects.UnaryPredicate;

import java.util.AbstractMap;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

/**
 * Please ignore some Warnings...
 */
public class NonModifyingSequence {
	public static boolean anyOf(Collection c, UnaryPredicate f) {
		for (Object o : c) {
			if (f.call(o))
				return true;
		}
		return false;
	}

	public static boolean allOf(Collection c, UnaryPredicate f) {
		for (Object o : c) {
			if (!f.call(o))
				return false;
		}
		return true;
	}

	public static boolean noneOf(Collection c, UnaryPredicate f) {
		for (Object o : c) {
			if (f.call(o))
				return false;
		}
		return true;
	}

	public static void forEach(Collection c, UnaryOp f) {
		for (Object o : c)
			f.call(o);
	}

	public static int count_if(Collection c, UnaryPredicate f) {
		int num = 0;
		for (Object o : c) {
			if (f.call(o))
				num++;
		}
		return num;
	}

	public static int count(Collection c, Object match) {
		int num = 0;
		for (Object o : c) {
			if (o.equals(match))
				num++;
		}
		return num;
	}

	// Unlike other languages, we can't get mismatch data with this mismatch
	// method.
	public static Pair mismatch(Iterable a, Iterable b, BinaryPredicate f) {
		Iterator it1 = a.iterator();
		Iterator it2 = b.iterator();
		while (it1.hasNext() && it2.hasNext()) {
			if (!f.call(it1.next(), it2.next()))
				return new Pair(it1, it2);
		}
		return null;
	}

	public static boolean equal(Iterable a, Iterable b, BinaryPredicate f) {
		Iterator it1 = a.iterator();
		Iterator it2 = b.iterator();
		while (it1.hasNext() && it2.hasNext()) {
			if (!f.call(it1.next(), it2.next()))
				return false;
		}
		if (it1.hasNext() || it2.hasNext())
			return false;
		return true;
	}

	@SuppressWarnings("unchecked")
	public static <T extends Object> T find(Collection c, T match) {
		for (Object o : c) {
			if (o.equals(match))
				return (T)o;
		}
		return null;
	}

	@SuppressWarnings("unchecked")
	public static <T extends Object> T findIf(Collection c, UnaryPredicate f) {
		for (Object o : c) {
			if (f.call(o))
				return (T)o;
		}
		return null;
	}

	@SuppressWarnings("unchecked")
	public static <T extends Object> T findIfNot(Collection c, UnaryPredicate f) {
		for (Object o : c) {
			if (!f.call(o))
				return (T)o;
		}
		return null;
	}

	public static Iterator findFirstOf(Collection a, Collection b, BinaryPredicate f) {
		for (Iterator i = a.iterator(); i.hasNext();) {
			Object o = i.next();
			for (Object j : b) {
				if (f.call(o, j))
					return i;
			}
		}
		return null;
	}

	// iterator in Java can't be cloned, so I just implement a List version
	public static int findEnd(List a, List b, BinaryPredicate f) {
		int result = -1;
		int aSize = a.size(), bSize = b.size();
		for (int i = 0; i < aSize; i++) {
			int start = i;
			for (Object j : b) {
				if (start >= aSize || !f.call(a.get(start), j)) {
					break;
				}
				start++;
			}
			if (start - i == bSize)
				result = i;
		}
		return result;
	}

	public static Iterator adjacent_find(Collection c, BinaryPredicate f) {
		if (c.size() < 2)
			return null;
		Iterator i = c.iterator();
		Object prev = i.next();
		while (i.hasNext()) {
			Object cur = i.next();
			if (f.call(prev, cur))
				return i;
			prev = cur;
		}
		return null;
	}

	// only implement the Array version of search*, so we can get rid of iterator
	public static int search(List a, List b, BinaryPredicate f) {
		int size = a.size();
		if (b.size() == 0)
			return 0;
		for (int i = 0, j = 0; i < size; i++) {
			if (!f.call(a.get(i), b.get(j))) {
				j = 0;
			}
			else {
				j++;
			}
			if (j == b.size())
				return i;
		}
		return size;
	}

	public static int searchN(List a, int count, Object match, BinaryPredicate f) {
		int size = a.size();
		if (count <= 0)
			return 0;
		for (int i = 0, j = 0; i < size; i++) {
			if (!f.call(a.get(i), match)) {
				j = 0;
			}
			else {
				j++;
			}
			if (j == count)
				return i;
		}
		return size;
	}
}
